### ISSUE TITLE
Modified Model artifacts should be generated and deployed 

### ISSUE THREAD

**Issue by rashiatmarklogic** [marklogic/entity-services#227]
So, the issue is : Lets say I have model1 and now I am changing the model as model2. I can  run `version-translator-generate()` to generate a module which will convert the source instances into converted instances, but, do I need to remove all artifacts which were generated from model1 and deploy the new artifacts generated from model2 ?

If yes, then we should add this as part of the recommendations somewhere in documentation. This can be source of confusion for people.

**Comment by rashiatmarklogic** [comment]
@bsrikan keeping you in loop too.

**Comment by rashiatmarklogic** [comment]
part of the experiment is also - to see what happens when the artifatcs from model1 and model2 coexist. According to Charles the only artifact which will create some mess is "schema-generate" else all other artifacts can be modified for model2.

**Comment by grechaw** [comment]
I would refine what you've said slightly.  The presence or absence of the old artifacts has nothing technically to do with Entity Services.  Entity Services doesn't really proscribe where to store documents or how to manage your data.

It is however true that the generated "application scaffold" artifacts do not make an attempt to specify versions except in comments.  A developer does need to be aware of what code and configuration she is using in a deployment, and we would do well to warn Entity Services users of the risk of conflicting, say, schemas.

**Comment by grechaw** [comment]
I'll take the action to put some information about versions and artifacts in the functional specification.

**Comment by grechaw** [comment]
Hi @kcoleman-marklogic , I've added you as also assigned to this issue.  In the ongoing work on the versions example, we see that, although one *can* insert schemas or extraction templates for multiple versions of a document at once, somebody in charge of a data migration must manage these assets with care.

Extraction templates are generated such that they only match instances of a given version.  Therefore templates will not over-index instance data.  However, it helps to use distinct schema names for different versions.  A convention for 'latest' might help with managing extraction tamplates.

I have not explored the ramifactions of schemas yet.  It is entirely possible to choose which schema is in context for which validation scenario.  I'm not as clear how types are resolved given potentially conflicting schemas.

Database configuration is of course a particular state.  The administrator has to manage the addition and removal of indexes throughout the life of the data hub.

I assigned you because, in order to resolve this issue, I have to work this information into an example README and it should also be covered in the entity services guide?

**Comment by grechaw** [comment]
I put in a bit of documentation about this issue into my latest PR, for the versions example.  It explains how TDE templates are best scoped to just a single version, and suggests how one might want to manage the transition of models.

**Comment by bsrikan** [comment]
Information on how the assets should be handled in various data migration scenarios is detailed. Looks good to me. Passing the issue on to @kcoleman-marklogic to cover this in entity services guide.

**Comment by kcoleman-marklogic** [comment]
@grechaw Did you give more thought to this:

> I have not explored the ramifactions of schemas yet. It is entirely possible to choose which schema is in context for which validation scenario. I'm not as clear how types are resolved given potentially conflicting schemas.

It seems to me that you simply cannot install (or at least use) multiple schemas. You could modify the complex type names but that doesn't help with the entity element definition. That is, I could fiddle with this to make the type name "PersonType-1.0" or some such thing:
```
<xs:complexType name="PersonType" mixed="true">...</xs:complexType>
```
But what do I do with this guy?
```
<xs:element name="Person" type="PersonType"/>
```
I poked around a bit in the XML Schema spec, and see you might be able to do something with a union, but that looks like major surgery on the generated schemas and most assuredly not something I have time to figure out how to do.

TL;DR: What guidance should I give wrt schemas?

**Comment by kcoleman-marklogic** [comment]
Maybe I figured it out, sorta. You can do this in XQuery, at least if you're doing explicit validation:
```
import schema default element namespace "" at "/es-gs/person-0.0.2.xsd";
```
No clue if you can do likewise in JS. I thought you might also be able to pull something off with `@xsi:schemaLocation` on the instance root node, but I can't seem to get that work.

**Comment by kcoleman-marklogic** [comment]
One more question for you @grechaw then I'll leave you peace about versioning for the moment. You said:

`Extraction templates are generated such that they only match instances of a given version. Therefore templates will not over-index instance data. However, it helps to use distinct schema names for different versions. A convention for 'latest' might help with managing extraction tamplates.`

I do not understand what you mean in the last two sentences. What schema is applicable to the template, and what kind of "convention for latest" are you referring to?

**Comment by grechaw** [comment]
It's a fair question, and not a fully developed answer.

Just as we have a URL for the REST API, /LATEST/documents  , one could devise (and maybe we should devise) a convention where one would create a duplicate TDE template, with a schema name that incorporates "latest" somehow.

Say, version 0.0.1 is Model.Type.  Also present is a duplicate called ModelLatest.Type
The next version is introduced and the ModelLatest schema is updated to be the latest version.
Then any client who is flexible enough to look for the most current version could always use ModelLatest schema.

That's what I meant anyhow, it might not be worth developing this idea further right now.

**Comment by bsrikan** [comment]
@kcoleman-marklogic This is regarding the last 2 sentences you were referring to.

Say you have a versioning scenario and working with artifacts generated for both versions of the model(model1 and model2 say).So you have 2 sets each of tde, envelopes, xsd, options etc. If you look at the tde, ver1 and ver2, you would find that the schema name is same as the title of the descriptor. 

This is all OK, but the problem Charles and I encountered was that sql queries returned rows that were duplicate and some had "null". This was because both tde views were at work. So when schema name was edited to make them distinct, the duplicate and null issue was gone. 

Convention for 'latest' is probably an idea to developers to append to schema name "_latest" maybe when generating extraction templates in the versioning scenario.
Something like what @grechaw did for #214 maybe.

**Comment by kcoleman-marklogic** [comment]
Thank you both. I finally figured out you guys are probably talking about the template/rows/row/schema-name, not the schema you can generate with ES. I'll read over all this again with that insight and see if it starts making more sense to me. :)

**Comment by grechaw** [comment]
Hah, yep, the TDE schema, not XSD schema.  MarkLogic glossary/ontology anyone?

As far as XSD goes, you have probably learned better than I how to do this by now. 
Mary confirmed that one can manage schema location as you have above, so if people want to monkey around with having multiple versions of schemas available for validating, there's a known method to do that.

**Comment by kcoleman-marklogic** [comment]
Not sure what the expectation was when this got assigned to me, but I've said all I'm going to say about versioning for now, so I am going to ship this bug. If there are versioning related doc issues, they should be reported as doc bugs (preferably on bugtrack).
